cache-prefetch.go

```diff
@@ -0,0 +1,183 @@
+package rdns
+
+import (
+	"errors"
+	"expvar"
+	"github.com/miekg/dns"
+	"github.com/sirupsen/logrus"
+	"sync"
+	"time"
+)
+
+type CachePrefetch struct {
+	CachePrefetchOptions
+	id       string
+	resolver Resolver
+	mu       sync.Mutex
+	metrics  *CachePrefetchMetrics
+}
+
+
+type CachePrefetchMetrics struct {
+	// Cache hit count.
+	domainEntries map[string]CachePrefetchEntry
+}
+type CachePrefetchEntry struct {
+	// request hit count.
+	hit 			  expvar.Int
+
+	// 0 for no prefetching job
+	// 1 for prefetching job is active
+	// 2 for stopped by error
+	prefetchingStatus expvar.Int
+	// store the time to live
+	ttl				  expvar.Int
+	// fetching error count for discarding error prone fetches
+	errorCount 		  expvar.Int
+}
+
+var _ Resolver = &CachePrefetch{}
+
+type CachePrefetchOptions struct {
+	//// Time of cache record ttl polling for record prefetch
+	CacheTTLPollingCheckInterval time.Duration
+	//// Min record time remaining check for expire
+	MinRecordTimeRemainingPercent uint64
+
+	// Number of hits a record gets before prefetch on a record is started
+	RecordQueryHitsMin int64
+	// Max number of responses to check in the cache. Defaults to 0 which means no limit. If
+	// the limit is reached, the least-recently used entry is removed from the cache.
+	//TODO
+	//RecordCheckCapacity uint32
+	// Cache to check records from
+	//CacheResolver Cache
+	// Allows control over the order of answer RRs in cached responses. Default is to keep
+	// the order if nil.
+	// TODO
+	//ShuffleAnswerFunc AnswerShuffleFunc
+}
+
+func NewCachePrefetch(id string, resolver Resolver, opt CachePrefetchOptions) *CachePrefetch {
+	c := &CachePrefetch{
+		CachePrefetchOptions: opt,
+		id:                   id,
+		resolver:             resolver,
+		metrics: &CachePrefetchMetrics{
+			domainEntries: map[string]CachePrefetchEntry{},
+		},
+	}
+	if c.CacheTTLPollingCheckInterval == 0 {
+		c.CacheTTLPollingCheckInterval = time.Minute
+	}
+	if c.RecordQueryHitsMin == 0 || c.RecordQueryHitsMin == 1 || c.RecordQueryHitsMin == -1 {
+		// Set to hit after one record hit
+		// fetch opportunistically
+		c.RecordQueryHitsMin = 1
+	}
+	return c
+}
+
+func (r *CachePrefetch) Resolve(q *dns.Msg, ci ClientInfo) (*dns.Msg, error) {
+	if len(q.Question) < 1 {
+		return nil, errors.New("no question in query")
+	}
+	// While multiple questions in one DNS message is part of the standard,
+	// it's not actually supported by servers. If we do get one of those,
+	// just pass it through and bypass caching.
+	if len(q.Question) > 1 {
+		return r.resolver.Resolve(q, ci)
+	}
+
+	go r.requestAddPrefetchJob(q, ci)
```

Comment:
It's not a good idea to start a goroutine per query, that can easily overload the server. This can instead be done in a single goroutine or keep it synchronous.

Reply:
Its been done in a single goroutine